### ISSUE TITLE
Creating an UmAlQuraCalendar fails on Linux

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#6774]
## Problem
In PR #6733 the following error occured when constructing an UmAlQuaraCalendar on Linux. The Windows build and test did not fail or throw any exceptions. They passed as expected.

> 15:52:30 ---- System.NullReferenceException : Object reference not set to an instance of an object.
15:52:30 Stack Trace:
15:52:30 at System.Globalization.UmAlQuraCalendar.get_MaxSupportedDateTime()
15:52:30 at System.Globalization.Tests.CalendarHelpers.MaxCalendarYearInEra(Calendar calendar, Int32 era)
15:52:30 at System.Globalization.Tests.CalendarHelpers.ToDateTime_Invalid(Calendar calendar)
15:52:30 ----- Inner Stack Trace -----
15:52:30 at System.Globalization.UmAlQuraCalendar.InitDateMapping()
15:52:30 at System.Globalization.UmAlQuraCalendar..cctor()
15:52:30 System.Globalization.Tests.CalendarHelpers.GetMonth_Invalid [FAIL]
15:52:30 System.TypeInitializationException : The type initializer for 'System.Globalization.UmAlQuraCalendar' threw an exception.
15:52:30 ---- System.NullReferenceException : Object reference not set to an instance of an object.
15:52:30 Stack Trace:
15:52:30 at System.Globalization.UmAlQuraCalendar.get_MinSupportedDateTime()
15:52:30 at System.Globalization.Tests.CalendarHelpers.d__10.MoveNext()
15:52:30 at System.Linq.Enumerable.SelectEnumerableIterator2.MoveNext() 15:52:30 ----- Inner Stack Trace ----- 15:52:30 at System.Globalization.UmAlQuraCalendar.InitDateMapping() 15:52:30 at System.Globalization.UmAlQuraCalendar..cctor() 15:52:30 System.Globalization.Tests.CalendarHelpers.IsLeapMonth_Invalid [FAIL] 15:52:30 System.TypeInitializationException : The type initializer for 'System.Globalization.UmAlQuraCalendar' threw an exception. 15:52:30 ---- System.NullReferenceException : Object reference not set to an instance of an object. 15:52:30 Stack Trace: 15:52:30 at System.Globalization.UmAlQuraCalendar.get_MaxSupportedDateTime() 15:52:30 at System.Globalization.Tests.CalendarHelpers.MaxCalendarYearInEra(Calendar calendar, Int32 era) 15:52:30 at System.Globalization.Tests.CalendarHelpers.<Year_Month_Day_Era_TestData>d__9.MoveNext() 15:52:30 at System.Linq.Enumerable.SelectEnumerableIterator2.MoveNext()
15:52:30 ----- Inner Stack Trace -----
15:52:30 at System.Globalization.UmAlQuraCalendar.InitDateMapping()
15:52:30 at System.Globalization.UmAlQuraCalendar..cctor()
15:52:30 Finished: System.Globalization.Calendars.Tests

## Investigation
It seems like this exception is thrown in [this](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Globalization/UmAlQuraCalendar.cs#L42) method (`InitDateMapping`)

There is a comment in the code:
```
//Direct inline initialization of DateMapping array would produce a lot of code bloat.
// We take advantage of C# compiler compiles inline initialization of primitive type array into very compact code.
// So we start with raw data stored in primitive type array, and initialize the DateMapping out of it
```
Maybe this behaviour cannot be guaranteed on roslyn on Linux?

/cc @stephentoub @tarekgh

**Comment by tarekgh** [comment]
Thanks @hughbe. I'll try to take a look at this later.

**Comment by tarekgh** [comment]
@jkotas could you have a look at this issue? 

UnALQuraCalendar class is defining the following on order:

        static readonly DateMapping[] HijriYearInfo = InitDateMapping();

        static readonly short[] rawData = new short[] { ... }

        static DateMapping[] InitDateMapping()
        {
            DateMapping[] mapping = new DateMapping[rawData.Length / 4];
            ....
        }

On Linux it looks the static initialization done in exact same order, so InitDateMapping will get called before rawData get initialized and this will cause null reference exception. if we move the line static readonly short[] rawData = new short[] { ... } to the top, it'll work. looks we have something wrong on static initialization logic on Linux.

**Comment by jkotas** [comment]
Windows and Unix use different implementations of globalization.

src\mscorlib\src\System\Globalization\*  is used for Windows
src\mscorlib\corefx\System\Globalization\* is used for Unix - it is the refactored impl from System.Private.CoreLib

The initialization code differs between the two, and the Unix one has a bug ... no magic. 

(Ideally, we would switch CoreCLR on Windows to use the refactored impl too at some point... .)

**Comment by tarekgh** [comment]
@jkotas why this implementation is not failing in PN? 
https://github.com/tarekgh/corert/blob/master/src/System.Private.CoreLib/src/System/Globalization/UmAlQuraCalendar.cs#L15

**Comment by jkotas** [comment]
Because PN tries to do magic to run static constructor at compile time - this optimization hides the bug.

(Note that this optimization is not CoreRT today, and the bug would still repro in CoreRT.)

**Comment by tarekgh** [comment]
so we have magic somewhere :smile: 
will fix that for coreclr and corert

**Comment by hughbe** [comment]
Fixed by dotnet/coreclr#3695
